bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -2525,185 +2570,206 @@ void BCStateTran::processData(bool lastInBatch) {
-          g.txn()->setResPage(e->pageId, e->checkpointNumber, e->pageDigest, e->page);
-          LOG_DEBUG(logger_, "Update page " << e->pageId);
-        }
-      }
-      ConcordAssert(g.txn()->hasCheckpointBeingFetched());
-
-      DataStore::CheckpointDesc cp = g.txn()->getCheckpointBeingFetched();
-
-      // set stored data
-      ConcordAssertEQ(g.txn()->getFirstRequiredBlock(), 0);
-      ConcordAssertEQ(g.txn()->getLastRequiredBlock(), 0);
-      ConcordAssertGT(cp.checkpointNum, g.txn()->getLastStoredCheckpoint());
-      ConcordAssert(ioContexts_.empty());
-
-      g.txn()->setCheckpointDesc(cp.checkpointNum, cp);
-      g.txn()->deleteCheckpointBeingFetched();
-      deleteOldCheckpoints(cp.checkpointNum, g.txn());
-      nextRequiredBlock_ = 0;
-      digestOfNextRequiredBlock.makeZero();
-      clearAllPendingItemsData();
-
-      // Metrics set at the end of the block to prevent transaction abort from
-      // leaving inconsistencies.
-      metrics_.last_stored_checkpoint_.Get().Set(cp.checkpointNum);
-      metrics_.checkpoint_being_fetched_.Get().Set(0);
-
-      checkConsistency(config_.pedanticChecks);
-      if (!config_.isReadOnly && cre_) {
-        // At this point, we, if are not going to have another blocks in state transfer. So, we can safely stop CRE.
-        // if there is a reconfiguration state change that prevents us from starting another state transfer (i.e.
-        // scaling) then CRE probably won't work as well.
-        // 1. First, make sure we handled the most recent available updates.
-        concord::client::reconfiguration::PollBasedStateClient *pbc =
-            (concord::client::reconfiguration::PollBasedStateClient *)(cre_->getStateClient());
-        bool succ = false;
-        while (!succ) {
-          auto latestHandledUpdate = cre_->getLatestKnownUpdateBlock();
-          auto latestReconfUpdates = pbc->getStateUpdate(succ);
-          if (!succ) {
-            LOG_ERROR(logger_, "unable to get the latest reconfiguration updates");
-          }
-          for (const auto &update : latestReconfUpdates) {
-            if (update.blockid > latestHandledUpdate) {
-              succ = false;
-              break;
+        // Metrics set at the end of the block to prevent transaction abort from
+        // leaving inconsistencies.
+        metrics_.last_stored_checkpoint_.Get().Set(cp.checkpointNum);
+        metrics_.checkpoint_being_fetched_.Get().Set(0);
+
+        checkConsistency(config_.pedanticChecks);
+        if (!config_.isReadOnly && cre_) {
+          // At this point, we, if are not going to have another blocks in state transfer. So, we can safely stop CRE.
+          // if there is a reconfiguration state change that prevents us from starting another state transfer (i.e.
+          // scaling) then CRE probably won't work as well.
+          // 1. First, make sure we handled the most recent available updates.
```

Comment:
ProcessData() needs to be refactored sometime in future.

Reply:
I stringly agree, but It's not the time to refactor.
Luckily we are in dev branch.
We can do it later...for sure.